Added dark-mode diverging colormaps

martinosorb opened the pull request:
## PR summary
There is currently no dark-mode diverging colormaps. 
This adds three such colormaps, `berlin`, `managua`, and `vanimo`, which have been developed by @Crameri.

Rightfully, there is a resistance against adding new colormaps all the time, however:
- In issue #18608, there seemed to be a consensus that this would be a sufficiently different type of colormap to justify an addition.
- They have been developed according to scientific principles, and are meant to be perceptually uniform. The plot in the documentation (reproduced below) shows a better perceptual uniformity than many of the ones we currently have in matplotlib. The author's work principles are illustrated in [his paper](https://www.nature.com/articles/s41467-020-19160-7).
- As requested by the author in the Issue discussion, I have kept the same names, and included a reference to the DOI of his work.
- The "Choosing Colormaps in Matplotlib" user guide has been updated:
![diverging](https://github.com/user-attachments/assets/6193fc84-3008-491d-8d03-3ee5a4c86bd2)

Closes #18608.



## PR checklist
<!-- Please mark any checkboxes that do not apply to this PR as [N/A].-->

- [X] "closes #0000" is in the body of the PR description to [link the related issue](https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue)
- [N/A] new and changed code is [tested](https://matplotlib.org/devdocs/devel/testing.html)
- [X] *Plotting related* features are demonstrated in an [example](https://matplotlib.org/devdocs/devel/document.html#write-examples-and-tutorials)
- [X] *New Features* and *API Changes* are noted with a [directive and release note](https://matplotlib.org/devdocs/devel/api_changes.html#announce-changes-deprecations-and-new-features)
- [N/A] Documentation complies with [general](https://matplotlib.org/devdocs/devel/document.html#write-rest-pages) and [docstring](https://matplotlib.org/devdocs/devel/document.html#write-docstrings) guidelines

<!--We understand that PRs can sometimes be overwhelming, especially as the
reviews start coming in.  Please let us know if the reviews are unclear or
the recommended next step seems overly demanding, if you would like help in
addressing a reviewer's comments, or if you have been waiting too long to hear
back on your PR.-->

jklymak commented:
From the circle log
```
/home/circleci/project/doc/users/next_whats_new/diverging_colormaps.rst:4: WARNING: py:obj reference target not found: berlin
/home/circleci/project/doc/users/next_whats_new/diverging_colormaps.rst:4: WARNING: py:obj reference target not found: managua
/home/circleci/project/doc/users/next_whats_new/diverging_colormaps.rst:4: WARNING: py:obj reference target not found: vanimo
```

I think you just want those as quotes, not back quotes?

jklymak commented on lib/matplotlib/_cm_listed.py:
```diff
@@ -2057,6 +2057,779 @@
+    [0.71945, 0.9592, 0.60112],
+    [0.72782, 0.96989, 0.61646],
+    [0.7362, 0.98063, 0.63191],
+    [0.74458, 0.99141, 0.64748]]
```
This sees fine, but are these really defined by a colourable, or do they also have a linear segmented definition?  If so, that would be much more compact....

jklymak replied:
Hmm, I guess not, according to the originals...

martinosorb replied:
I don't know enough to dare changing from the original version, personally.

jklymak reviewed (commented):
Docs need fixing, but otherwise looks OK to me.  

Bonus points if you add a section to Choosing Colormaps to get more exposure for this.  Not that I agree with everything in Choosing Colormaps, but...

martinosorb commented:
Thank you for the quick reply.
Ok, I fixed the quotes, let's see if this fixes it.

Regarding Choosing Colormaps, I had added a paragraph saying 
```
  # Berlin, Managua, and Vanimo are dark-mode diverging colormaps, with minimum
  # lightness at the center, and maximum at the extremes. These are taken from
  # F. Crameri's [scientific colour maps]_ version 8.0.1.
```

Is that what you meant or did you want something more?

jklymak commented:
> Is that what you meant or did you want something more?

I was suggesting adding to:

https://matplotlib.org/stable/users/explain/colors/colormaps.html

which is one of our highest ranking pages for google search.

jklymak commented:
Oh, I'm sorry - I completely spaced that `users/explain` was under galleries now, despite moving it there myself!  Anyhow, looks great!

martinosorb commented on doc/users/next_whats_new/diverging_colormaps.rst:
```diff
@@ -0,0 +1,24 @@
+
+.. plot::
+    :include-source: true
+    :alt: Example of creating 4 horizontal boxplots.
```
My bad, sorry for forgetting that. I've fixed it now.

martinosorb commented:
Sorry for forgetting that. I've now fixed it.

jklymak commented on doc/users/next_whats_new/diverging_colormaps.rst:
```diff
@@ -0,0 +1,24 @@
+
+.. plot::
+    :include-source: true
+    :alt: Example figures using "imshow" with dark-mode diverging colormaps on positive and negative data. First panel: "berlin" (blue to red with a black center); second panel: "managua" (orange to cyan with a dark purple center); third panel: "vanimo" (pink to green with a black center).
```
We don't lint rst files, but pretty sure this should be wrapped.  I think as long as you indent the plot directive should still be OK.

martinosorb replied:
I wasn't sure whether breaking the line would send me out of the `alt` variable. I checked the previous what's new files and they seem to go on without breaking the line: https://raw.githubusercontent.com/matplotlib/matplotlib/ed56f95cb6f5b54525687dabbb0e660bafec4d94/doc/users/prev_whats_new/whats_new_3.9.0.rst

jklymak replied:
I'm not sure either, and maybe it was done that way on purpose.  But if not, then it would be nice to have wrapping

martinosorb replied:
Unfortunately I was unable to compile the docs locally due to some dependency issues, so I cannot check. Sorry.

jklymak replied:
Thats why we have CI

martinosorb replied:
That didn't work, I reverted.

jklymak reviewed (approved):
Thanks for your work on this!

jklymak commented:
Looks like the doc build is failing though, before we merge

martinosorb commented:
I rebased and squashed. CI is green.

jklymak commented:
Thanks @martinosorb !
